src/iam/certhandler/modules/certmodule.cpp

```diff
@@ -0,0 +1,418 @@
+    if (!templ) {
+        AOS_ERROR_WRAP(Error::Enum::eNoMemory);
+    }
+
+    templ->mSubject = subject;
+    templ->mDnsNames = mModuleConfig.mAlternativeNames;
+
+    StaticArray<crypto::asn1::ObjectIdentifier, crypto::cCertExtraExtCount> oids;
+
+    for (const auto& extKeyUsage : mModuleConfig.mExtendedKeyUsage) {
+        switch (extKeyUsage.GetValue()) {
+        case ExtendedKeyUsageEnum::eClientAuth:
+            oids.PushBack(cOidExtKeyUsageClientAuth);
+            break;
+
+        case ExtendedKeyUsageEnum::eServerAuth:
+            oids.PushBack(cOidExtKeyUsageServerAuth);
+            break;
+
+        default:
+            LOG_WRN() << "Unexpected extended key usage: type=" << GetCertType()
+                      << ", value=" << extKeyUsage.ToString();
+            break;
+        }
+    }
+
+    if (!oids.IsEmpty()) {
+        crypto::asn1::Extension ext;
+
+        ext.mId = cOidExtensionExtendedKeyUsage;
+        crypto::asn1::Encode(oids, ext.mValue);
+
+        auto err = templ->mExtraExtensions.PushBack(ext);
+        if (!err.IsNone()) {
+            return AOS_ERROR_WRAP(err);
+        }
+    }
+
+    return mX509Provider->CreateCSR(*templ, privKey, pemCSR);
+}
+
+Error CertModule::ApplyCert(const Array<uint8_t>& pemCert, CertInfo& info)
+{
+    auto certificates = MakeUnique<CertificateChain>(&mAllocator);
+    if (!certificates) {
+        return AOS_ERROR_WRAP(Error::Enum::eNoMemory);
+    }
+
+    auto err = mX509Provider->PEMToX509Certs(pemCert, *certificates);
+    if (!err.IsNone()) {
+        return err;
+    }
+
+    err = CheckCertChain(*certificates);
+    if (!err.IsNone()) {
+        return err;
+    }
+
+    StaticString<cPasswordLen> password;
+
+    err = mHSM->ApplyCert(*certificates, info, password);
+    if (!err.IsNone()) {
+        return err;
+    }
+
+    err = mStorage->AddCertInfo(GetCertType(), info);
+    if (!err.IsNone()) {
+        return err;
+    }
+
+    return TrimCerts(password);
+}
+
+Error CertModule::CreateSelfSignedCert(const String& password)
+{
+    auto key = CreateKey(password);
+    if (!key.mError.IsNone()) {
+        return key.mError;
+    }
+
+    const uint64_t serial = time::Time::Now().UnixNano();
+    auto           templ = MakeUnique<crypto::x509::Certificate>(&mAllocator);
+    if (!templ) {
+        return AOS_ERROR_WRAP(Error::Enum::eNoMemory);
+    }
+
+    templ->mSerial = Array<uint8_t>(reinterpret_cast<const uint8_t*>(&serial), sizeof(serial));
+    templ->mNotBefore = time::Time::Now();
+    templ->mNotAfter = time::Time::Now().Add(cValidSelfSignedCertPeriod);
+
+    auto err = mX509Provider->CreateDN("Aos Core", templ->mSubject);
+    if (!err.IsNone()) {
+        return err;
+    }
+
+    err = mX509Provider->CreateDN("Aos Core", templ->mIssuer);
+    if (!err.IsNone()) {
+        return err;
+    }
+
+    auto pemCert = MakeUnique<SelfSignedCertificate>(&mAllocator);
+    auto certInfo = MakeUnique<CertInfo>(&mAllocator);
```

Comment:
CertInfo is ok to keep in stack. Generally, your function should not exceed 2k of stack. It is rule for now.